# Data Cleaning Walkthrough

> Cleaning data in pythn

- author: Victor Omondi
- toc: true
- comments: true
- categories: [data-cleaning, python]
- image: images/dcp-shield.png

# Libraries

In [1]:
import re
import warnings

warnings.filterwarnings("ignore", message="numpy.ufunc size changed")

import numpy as np
import pandas as pd

import matplotlib.pyplot as plt
import seaborn as sns
%matplotlib inline
plt.style.use("ggplot")

In [2]:
pd.set_option("display.max_colwidth", None)
pd.set_option("display.max_columns", None)

# 1. Introduction

At many points in data science career, you'll need to be able to build complete, end-to-end data science projects . Data science projects usually consist of one of two things:

- An exploration and analysis of a set of data. One example might involve analyzing donors to political campaigns, creating a plot, and then sharing an analysis of the plot with others.
- An operational system that generates predictions based on data that updates continually. An algorithm that pulls in daily stock ticker data and predicts which stock prices will rise and fall would be one example.

You'll find the ability to create data science projects useful in several different contexts:

- Projects will helps build a portfolio, which is critical to finding a job as a data analyst or scientist.
- Working on projects will help you learn new skills and reinforce existing concepts.
- Most "real-world" data science and analysis work consists of developing internal projects.
- Projects allow you to investigate interesting phenomena and satisfy your curiosity.

We'll walk through the first part of a complete data science project, including how to acquire the raw data. The project will focus on exploring and analyzing a data set. We'll develop our data cleaning and storytelling skills, which will enable us to build complete projects on our own.

We'll focus primarily on data exploration. We'll also combine several messy data sets into a single clean one to make analysis easier.

## 1.1 Dataset

we'll be using data about New York City public schools, which can be found [here](https://data.cityofnewyork.us/browse?category=Education).

One of the most controversial issues in the U.S. educational system is the efficacy of standardized tests, and whether they're unfair to certain groups. Given our prior knowledge of this topic, investigating the correlations between SAT scores and demographics might be an interesting angle to take. We could correlate [SAT scores](https://en.wikipedia.org/wiki/SAT) with factors like race, gender, income, and more.

The SAT, or Scholastic Aptitude Test, is an exam that U.S. high school students take before applying to college. Colleges take the test scores into account when deciding who to admit, so it's fairly important to perform well on it.

The test consists of three sections, each of which has 800 possible points. The combined score is out of 2,400 possible points (while this number has changed a few times, the data set for our project is based on 2,400 total points). Organizations often rank high schools by their average SAT scores. The scores are also considered a measure of overall school district quality.

# 2. Finding All of the Relevant Data Sets

New York City makes its [data on high school SAT scores](https://data.cityofnewyork.us/Education/SAT-Results/f9bf-2cp4) available online, as well as the [demographics for each high school](https://data.cityofnewyork.us/Education/DOE-High-School-Directory-2014-2015/n3p6-zve2). The first few rows of the SAT data look like this:

![SAT scores](https://s3.amazonaws.com/dq-content/sat.png)

Unfortunately, combining both of the data sets won't give us all of the demographic information we want to use. We'll need to supplement our data with other sources to do our full analysis.

The same website has several related data sets covering demographic information and test scores. Here are the links to all of the data sets we'll be using:

* [SAT scores by school](https://data.cityofnewyork.us/Education/SAT-Results/f9bf-2cp4) - SAT scores for each high school in New York City
* [School attendance](https://data.cityofnewyork.us/Education/School-Attendance-and-Enrollment-Statistics-by-Dis/7z8d-msnt) - Attendance information for each school in New York City
* [Class size](https://data.cityofnewyork.us/Education/2010-2011-Class-Size-School-level-detail/urz7-pzb3) - Information on class size for each school
* [AP test results](https://data.cityofnewyork.us/Education/AP-College-Board-2010-School-Level-Results/itfs-ms3e) - Advanced Placement (AP) exam results for each high school (passing an optional AP exam in a particular subject can earn a student college credit in that subject)
* [Graduation outcomes](https://data.cityofnewyork.us/Education/Graduation-Outcomes-Classes-Of-2005-2010-School-Le/vh2h-md7a) - The percentage of students who graduated, and other outcome information
* [Demographics](https://data.cityofnewyork.us/Education/School-Demographics-and-Accountability-Snapshot-20/ihfw-zy9j) - Demographic information for each school
* [School survey](https://data.cityofnewyork.us/Education/NYC-School-Survey-2011/mnz3-dyi8) - Surveys of parents, teachers, and students at each school

All of these data sets are interrelated. We'll need to combine them into a single data set before we can find correlations.

# 3. Finding Background Information

Before we move into coding, we'll need to do some background research. A thorough understanding of the data will help us avoid costly mistakes, such as thinking that a column represents something other than what it does. Background research will also give us a better understanding of how to combine and analyze the data.

In this case, we'll want to research:

* [New York City](https://en.wikipedia.org/wiki/New_York_City)
* [The SAT](https://en.wikipedia.org/wiki/SAT)
* [Schools in New York City](https://en.wikipedia.org/wiki/List_of_high_schools_in_New_York_City)
* [Our data](https://data.cityofnewyork.us/browse?category=Education)

We can learn a few different things from these resources. For example:

* Only high school students take the SAT, so we'll want to focus on high schools.
* New York City is made up of five boroughs, which are essentially distinct regions.
* New York City schools fall within several different school districts, each of which can contains dozens of schools.
* Our data sets include several different types of schools. We'll need to clean them so that we can focus on high schools only.
* Each school in New York City has a unique code called a  `DBN` , or district borough number.
* Aggregating data by district will allow us to use the district mapping data to plot district-by-district differences.

# 4. Reading in the data

Once we've done our background research, we're ready to read in the data. For your convenience, we've placed all the data into the  `datasets`  folder. Here are all of the files in the folder:

* `ap_2010.csv`  - Data on [AP test results](https://data.cityofnewyork.us/Education/AP-College-Board-2010-School-Level-Results/itfs-ms3e)
* `class_size.csv`  - Data on [class size](https://data.cityofnewyork.us/Education/2010-2011-Class-Size-School-level-detail/urz7-pzb3)
* `demographics.csv`  - Data on [demographics](https://data.cityofnewyork.us/Education/School-Demographics-and-Accountability-Snapshot-20/ihfw-zy9j)
* `graduation.csv`  - Data on [graduation outcomes](https://data.cityofnewyork.us/Education/Graduation-Outcomes-Classes-Of-2005-2010-School-Le/vh2h-md7a)
* `hs_directory.csv`  - A directory of [high schools](https://data.cityofnewyork.us/Education/DOE-High-School-Directory-2014-2015/n3p6-zve2)
* `sat_results.csv`  - Data on [SAT scores](https://data.cityofnewyork.us/Education/SAT-Results/f9bf-2cp4)
* `survey_all.txt`  - Data on [surveys](https://data.cityofnewyork.us/Education/NYC-School-Survey-2011/mnz3-dyi8) from all schools
* `survey_d75.txt`  - Data on [surveys](https://data.cityofnewyork.us/Education/NYC-School-Survey-2011/mnz3-dyi8) from New York City [district 75](http://schools.nyc.gov/academics/specialEducation/D75/default.htm)

`survey_all.txt`  and  `survey_d75.txt`  are in more complicated formats than the other files. For now, we'll focus on reading in the  `CSV`  files only, and then explore them.

We'll read each file into a [pandas dataframe](http://pandas.pydata.org/pandas-docs/stable/generated/pandas.DataFrame.html), and then store all of the dataframes in a dictionary. This will give us a convenient way to store them, and a quick way to reference them later on.

In [3]:
data = {
    'ap_2010':pd.read_csv('datasets/ap_2010.csv'),
    'class_size':pd.read_csv('datasets/class_size.csv'),
    'demographics':pd.read_csv('datasets/demographics.csv'),
    'graduation':pd.read_csv('datasets/graduation.csv'),
    'hs_directory':pd.read_csv('datasets/hs_directory.csv'),
    'sat_results':pd.read_csv("datasets/sat_results.csv")
}

# 5. Exploring the SAT data

What we're mainly interested in is the SAT data set, which corresponds to the dictionary key  `sat_results` . This data set contains the SAT scores for each high school in New York City. We eventually want to correlate selected information from this data set with information in the other data sets.

Let's explore  `sat_results`  to see what we can discover. Exploring the dataframe will help us understand the structure of the data, and make it easier for us to analyze it.

In [4]:
data['sat_results'].head()

DBN                                    SCHOOL NAME  \
0  01M292  HENRY STREET SCHOOL FOR INTERNATIONAL STUDIES   
1  01M448            UNIVERSITY NEIGHBORHOOD HIGH SCHOOL   
2  01M450                     EAST SIDE COMMUNITY SCHOOL   
3  01M458                      FORSYTH SATELLITE ACADEMY   
4  01M509                        MARTA VALLE HIGH SCHOOL   

  Num of SAT Test Takers SAT Critical Reading Avg. Score SAT Math Avg. Score  \
0                     29                             355                 404   
1                     91                             383                 423   
2                     70                             377                 402   
3                      7                             414                 401   
4                     44                             390                 433   

  SAT Writing Avg. Score  
0                    363  
1                    366  
2                    370  
3                    359  
4                    384

# 6. Exploring The remaining data

We can make a few observations based on `sat_results` first five rows output:

* The  `DBN`  appears to be a unique ID for each school.
* We can tell from the first few rows of names that we only have data about high schools.
* There's only a single row for each high school, so each  `DBN`  is unique in the SAT data.
* We may eventually want to combine the three columns that contain SAT scores --  `SAT Critical Reading Avg. Score` ,  `SAT Math Avg. Score` , and  `SAT Writing Avg. Score`  -- into a single column to make the scores easier to analyze.

Given these observations, let's explore the other data sets to see if we can gain any insight into how to combine them.

In [5]:
for key in data.keys():
    display(data[key].head())

DBN                             SchoolName AP Test Takers   \
0  01M448           UNIVERSITY NEIGHBORHOOD H.S.              39   
1  01M450                 EAST SIDE COMMUNITY HS              19   
2  01M515                    LOWER EASTSIDE PREP              24   
3  01M539         NEW EXPLORATIONS SCI,TECH,MATH             255   
4  02M296  High School of Hospitality Management               s   

  Total Exams Taken Number of Exams with scores 3 4 or 5  
0                49                                   10  
1                21                                    s  
2                26                                   24  
3               377                                  191  
4                 s                                    s

CSD BOROUGH SCHOOL CODE                SCHOOL NAME GRADE  PROGRAM TYPE  \
0    1       M        M015  P.S. 015 Roberto Clemente     0K       GEN ED   
1    1       M        M015  P.S. 015 Roberto Clemente     0K          CTT   
2    1       M        M015  P.S. 015 Roberto Clemente     01       GEN ED   
3    1       M        M015  P.S. 015 Roberto Clemente     01          CTT   
4    1       M        M015  P.S. 015 Roberto Clemente     02       GEN ED   

  CORE SUBJECT (MS CORE and 9-12 ONLY) CORE COURSE (MS CORE and 9-12 ONLY)  \
0                                    -                                   -   
1                                    -                                   -   
2                                    -                                   -   
3                                    -                                   -   
4                                    -                                   -   

  SERVICE CATEGORY(K-9* ONLY)  NUMBER OF STUDENTS / SEATS FILLED  \
0                           -                               19.0   
1                           -                               21.0   
2                           -                               17.0   
3                           -                               17.0   
4                           -                               15.0   

   NUMBER OF SECTIONS  AVERAGE CLASS SIZE  SIZE OF SMALLEST CLASS  \
0                 1.0                19.0                    19.0   
1                 1.0                21.0                    21.0   
2                 1.0                17.0                    17.0   
3                 1.0                17.0                    17.0   
4                 1.0                15.0                    15.0   

   SIZE OF LARGEST CLASS DATA SOURCE  SCHOOLWIDE PUPIL-TEACHER RATIO  
0                   19.0         ATS                             NaN  
1                   21.0         ATS                             NaN  
2                   17.0         ATS                             NaN  
3                   17.0         ATS                             NaN  
4                   15.0         ATS                             NaN

DBN                       Name  schoolyear fl_percent  frl_percent  \
0  01M015  P.S. 015 ROBERTO CLEMENTE    20052006       89.4          NaN   
1  01M015  P.S. 015 ROBERTO CLEMENTE    20062007       89.4          NaN   
2  01M015  P.S. 015 ROBERTO CLEMENTE    20072008       89.4          NaN   
3  01M015  P.S. 015 ROBERTO CLEMENTE    20082009       89.4          NaN   
4  01M015  P.S. 015 ROBERTO CLEMENTE    20092010                    96.5   

   total_enrollment prek   k grade1 grade2 grade3 grade4 grade5 grade6 grade7  \
0               281   15  36     40     33     38     52     29     38    NaN   
1               243   15  29     39     38     34     42     46    NaN    NaN   
2               261   18  43     39     36     38     47     40    NaN    NaN   
3               252   17  37     44     32     34     39     49    NaN    NaN   
4               208   16  40     28     32     30     24     38    NaN    NaN   

  grade8 grade9 grade10 grade11 grade12  ell_num  ell_percent  sped_num  \
0    NaN    NaN     NaN     NaN     NaN     36.0         12.8      57.0   
1    NaN    NaN     NaN     NaN     NaN     38.0         15.6      55.0   
2    NaN    NaN     NaN     NaN     NaN     52.0         19.9      60.0   
3    NaN    NaN     NaN     NaN     NaN     48.0         19.0      62.0   
4    NaN    NaN     NaN     NaN     NaN     40.0         19.2      46.0   

   sped_percent ctt_num selfcontained_num  asian_num  asian_per  black_num  \
0          20.3      25                 9         10        3.6         74   
1          22.6      19                15         18        7.4         68   
2          23.0      20                14         16        6.1         77   
3          24.6      21                17         16        6.3         75   
4          22.1      14                14         16        7.7         67   

   black_per  hispanic_num  hispanic_per  white_num  white_per  male_num  \
0       26.3           189          67.3          5        1.8     158.0   
1       28.0           153          63.0          4        1.6     140.0   
2       29.5           157          60.2          7        2.7     143.0   
3       29.8           149          59.1          7        2.8     149.0   
4       32.2           118          56.7          6        2.9     124.0   

   male_per  female_num  female_per  
0      56.2       123.0        43.8  
1      57.6       103.0        42.4  
2      54.8       118.0        45.2  
3      59.1       103.0        40.9  
4      59.6        84.0        40.4

Demographic     DBN                            School Name    Cohort  \
0  Total Cohort  01M292  HENRY STREET SCHOOL FOR INTERNATIONAL      2003   
1  Total Cohort  01M292  HENRY STREET SCHOOL FOR INTERNATIONAL      2004   
2  Total Cohort  01M292  HENRY STREET SCHOOL FOR INTERNATIONAL      2005   
3  Total Cohort  01M292  HENRY STREET SCHOOL FOR INTERNATIONAL      2006   
4  Total Cohort  01M292  HENRY STREET SCHOOL FOR INTERNATIONAL  2006 Aug   

   Total Cohort Total Grads - n Total Grads - % of cohort Total Regents - n  \
0             5               s                         s                 s   
1            55              37                     67.3%                17   
2            64              43                     67.2%                27   
3            78              43                     55.1%                36   
4            78              44                     56.4%                37   

  Total Regents - % of cohort Total Regents - % of grads Advanced Regents - n  \
0                           s                          s                    s   
1                       30.9%                      45.9%                    0   
2                       42.2%                      62.8%                    0   
3                       46.2%                      83.7%                    0   
4                       47.4%                      84.1%                    0   

  Advanced Regents - % of cohort Advanced Regents - % of grads  \
0                              s                             s   
1                             0%                            0%   
2                             0%                            0%   
3                             0%                            0%   
4                             0%                            0%   

  Regents w/o Advanced - n Regents w/o Advanced - % of cohort  \
0                        s                                  s   
1                       17                              30.9%   
2                       27                              42.2%   
3                       36                              46.2%   
4                       37                              47.4%   

  Regents w/o Advanced - % of grads Local - n Local - % of cohort  \
0                                 s         s                   s   
1                             45.9%        20               36.4%   
2                             62.8%        16                 25%   
3                             83.7%         7                  9%   
4                             84.1%         7                  9%   

    Local - % of grads Still Enrolled - n Still Enrolled - % of cohort  \
0                    s                  s                            s   
1                54.1%                 15                        27.3%   
2  37.200000000000003%                  9                        14.1%   
3                16.3%                 16                        20.5%   
4                15.9%                 15                        19.2%   

  Dropped Out - n Dropped Out - % of cohort  
0               s                         s  
1               3                      5.5%  
2               9                     14.1%  
3              11                     14.1%  
4              11                     14.1%

dbn                                         school_name       boro  \
0  17K548                 Brooklyn School for Music & Theatre   Brooklyn   
1  09X543                    High School for Violin and Dance      Bronx   
2  09X327         Comprehensive Model School Project M.S. 327      Bronx   
3  02M280      Manhattan Early College School for Advertising  Manhattan   
4  28Q680  Queens Gateway to Health Sciences Secondary School     Queens   

  building_code    phone_number    fax_number grade_span_min  grade_span_max  \
0          K440    718-230-6250  718-230-6262              9              12   
1          X400    718-842-0687  718-589-9849              9              12   
2          X240    718-294-8111  718-294-8109              6              12   
3          M520  718-935-3477             NaN              9              10   
4          Q695    718-969-3155  718-969-3552              6              12   

  expgrade_span_min  expgrade_span_max  \
0               NaN                NaN   
1               NaN                NaN   
2               NaN                NaN   
3                 9               14.0   
4               NaN                NaN   

                                                  bus  \
0               B41, B43, B44-SBS, B45, B48, B49, B69   
1  Bx13, Bx15, Bx17, Bx21, Bx35, Bx4, Bx41, Bx4A, Bx6   
2   Bx1, Bx11, Bx13, Bx18, Bx2, Bx3, Bx32, Bx35, Bx36   
3                              M103, M15, M22, M5, M9   
4                                       Q25, Q46, Q65   

                                                                                                                                                            subway  \
0                                                                                                       2, 3, 4, 5, F, S to Botanic Garden ; B, Q to Prospect Park   
1                                                                                                                                            2, 5 to Intervale Ave   
2                                                                                                                              4 to Mt Eden Ave ; B, D to 170th St   
3  1 to Chambers St ; 2, 3, A, C to Fulton St-Broadway-Nassau ; 4, 5, 6, J, Z to Brooklyn Bridge - City Hall ; E to Park Place ; N, Q to Canal St ; R to City Hall   
4                                                                                                                                                              NaN   

   primary_address_line_1      city state_code    zip  \
0      883 Classon Avenue  Brooklyn         NY  11225   
1        1110 Boston Road     Bronx         NY  10456   
2      1501 Jerome Avenue     Bronx         NY  10452   
3        411 Pearl Street  New York         NY  10038   
4  160-20 Goethals Avenue   Jamaica         NY  11432   

                                        website  total_students  \
0                           Bkmusicntheatre.com           399.0   
1                                  www.hsvd.org           378.0   
2  http://schools.nyc.gov/schoolportals/09/X327           543.0   
3                                           NaN             NaN   
4                         www.queensgateway.com           811.0   

                           campus_name                  school_type  \
0  Prospect Heights Educational Campus                          NaN   
1            Morris Educational Campus                          NaN   
2  DOE New Settlement Community Campus                          NaN   
3   Murry Bergtraum Educational Campus  NYC P-Tech 9-14, New School   
4                                  NaN                          NaN   

                                                                                                                                                                                                                                                                                                              

DBN                                    SCHOOL NAME  \
0  01M292  HENRY STREET SCHOOL FOR INTERNATIONAL STUDIES   
1  01M448            UNIVERSITY NEIGHBORHOOD HIGH SCHOOL   
2  01M450                     EAST SIDE COMMUNITY SCHOOL   
3  01M458                      FORSYTH SATELLITE ACADEMY   
4  01M509                        MARTA VALLE HIGH SCHOOL   

  Num of SAT Test Takers SAT Critical Reading Avg. Score SAT Math Avg. Score  \
0                     29                             355                 404   
1                     91                             383                 423   
2                     70                             377                 402   
3                      7                             414                 401   
4                     44                             390                 433   

  SAT Writing Avg. Score  
0                    363  
1                    366  
2                    370  
3                    359  
4                    384

# 7. Reading in the survey data

We can make some observations based on the first few rows of each dataframe.
* Each data set appears to either have a  `DBN`  column, or the information we need to create one. That means we can use a  `DBN`  column to combine the data sets. First we'll pinpoint matching rows from different data sets by looking for identical  `DBN` s, then group all of their columns together in a single data set.
* Some fields look interesting for mapping -- particularly  `Location 1` , which contains coordinates inside a larger string.
* Some of the data sets appear to contain multiple rows for each school (because the rows have duplicate  `DBN`  values). That means we’ll have to do some preprocessing to ensure that each  `DBN`  is unique within each data set. If we don't do this, we'll run into problems when we combine the data sets, because we might be merging two rows in one data set with one row in another data set.

Before we proceed with the merge, we should make sure we have all of the data we want to unify. We mentioned the survey data earlier ( `survey_all.txt`  and  `survey_d75.txt` ), but we didn't read those files in because they're in a slightly more complex format.

Each survey text file looks like this:

```
dbn bn  schoolname  d75 studentssurveyed    highschool  schooltype  rr_s
"01M015"    "M015"  "P.S. 015 Roberto Clemente" 0   "No"    0   "Elementary School"     88
```

The files are  *tab delimited*  and encoded with  `Windows-1252`  encoding. An encoding defines how a computer stores the contents of a file in binary. The most common encodings are  `UTF-8`  and  `ASCII` .  `Windows-1252`  is rarely used, and can cause errors if we read such a file in without specifying the encoding. If you'd like to read more about encodings, [here's](http://kunststube.net/encoding/) a good primer.

We'll need to specify the encoding and delimiter to the pandas [pandas.read_csv()](http://pandas.pydata.org/pandas-docs/stable/generated/pandas.read_csv.html) function to ensure it reads the surveys in properly.

After we read in the survey data, we'll want to combine it into a single dataframe. We can do this by calling the [pandas.concat()](http://pandas.pydata.org/pandas-docs/stable/generated/pandas.concat.html) function

In [6]:
all_survey = pd.read_csv("datasets/survey_all.txt", delimiter="\t", encoding="windows-1252")
d75_survey = pd.read_csv("datasets/survey_d75.txt", delimiter="\t", encoding="windows-1252")
survey = pd.concat([all_survey, d75_survey], axis=0)
survey.head()

dbn    bn                      schoolname  d75 studentssurveyed  \
0  01M015  M015       P.S. 015 Roberto Clemente    0               No   
1  01M019  M019             P.S. 019 Asher Levy    0               No   
2  01M020  M020            P.S. 020 Anna Silver    0               No   
3  01M034  M034  P.S. 034 Franklin D. Roosevelt    0              Yes   
4  01M063  M063       P.S. 063 William McKinley    0               No   

   highschool                  schooltype  rr_s  rr_t  rr_p    N_s   N_t  \
0         0.0           Elementary School   NaN    88    60    NaN  22.0   
1         0.0           Elementary School   NaN   100    60    NaN  34.0   
2         0.0           Elementary School   NaN    88    73    NaN  42.0   
3         0.0  Elementary / Middle School  89.0    73    50  145.0  29.0   
4         0.0           Elementary School   NaN   100    60    NaN  23.0   

     N_p  nr_s  nr_t  nr_p  saf_p_11  com_p_11  eng_p_11  aca_p_11  saf_t_11  \
0   90.0     0    25   150       8.5       7.6       7.5       7.8       7.5   
1  161.0     0    33   269       8.4       7.6       7.6       7.8       8.6   
2  367.0     0    48   505       8.9       8.3       8.3       8.6       7.6   
3  151.0   163    40   301       8.8       8.2       8.0       8.5       7.0   
4   90.0     0    23   151       8.7       7.9       8.1       7.9       8.4   

   com_t_11  eng_t_11  aca_t_11  saf_s_11  com_s_11  eng_s_11  aca_s_11  \
0       7.8       7.6       7.9       NaN       NaN       NaN       NaN   
1       8.5       8.9       9.1       NaN       NaN       NaN       NaN   
2       6.3       6.8       7.5       NaN       NaN       NaN       NaN   
3       6.2       6.8       7.8       6.2       5.9       6.5       7.4   
4       7.3       7.8       8.1       NaN       NaN       NaN       NaN   

   saf_tot_11  com_tot_11  eng_tot_11  aca_tot_11  p_q2h  p_q7a  p_q7b  p_q7c  \
0         8.0         7.7         7.5         7.9    8.0    8.2    8.3    7.5   
1         8.5         8.1         8.2         8.4    7.7    7.9    8.0    7.3   
2         8.2         7.3         7.5         8.0    8.1    8.8    8.9    8.5   
3         7.3         6.7         7.1         7.9    8.1    8.5    8.8    8.2   
4         8.5         7.6         7.9         8.0    8.0    8.4    8.6    8.0   

   p_q7d  p_q8a  p_q8b  p_q8c  p_q8d  p_q8e  p_q8f  p_q2b  p_q2d  p_q2e  \
0    7.9    6.8    8.7    9.7    8.7    9.9    9.9    7.7    8.3    7.9   
1    7.7    6.5    8.8    9.4    8.7   10.0    9.9    7.5    8.2    7.9   
2    8.4    7.6    9.2    9.4    9.2    9.8    9.7    8.4    8.8    8.5   
3    8.3    7.3    9.2    9.4    9.1    9.8    9.7    8.3    8.7    8.3   
4    8.0    6.5    8.8    9.6    9.4   10.0   10.0    7.6    8.5    8.0   

   p_q2f  p_q2g  p_q3a  p_q3b  p_q4b  p_q4c  p_q11c  p_q2a  p_q2c  p_q3c  \
0    8.1    7.5    7.3    6.7    7.6    7.9     7.5    8.0    7.4    8.7   
1    8.0    7.2    7.0    6.9    8.0    8.4     7.4    7.6    7.1    9.0   
2    8.3    8.0    7.6    7.3    8.6    8.7     8.4    8.8    8.2    8.8   
3    8.1    7.6    7.6    7.5    8.6    8.6     8.4    8.5    8.1    8.8   
4    8.2    7.4    7.8    7.2    8.1    7.9     7.8    8.2    8.3    9.2   

   p_q6a  p_q6b  p_q11d  p_q11e  p_q5  p_q4a  p_q4d  p_q4e  p_q11a  p_q11b  \
0    6.3    6.6     7.6     7.6   7.4    7.8    7.4    NaN     8.3     7.5   
1    6.4    6.5     7.6     7.8   8.6    7.7    7.8    NaN     8.3     7.6   
2    7.9    6.8     8.5     8.5   8.7    8.6    8.5    NaN     8.9     8.4   
3    7.3    6.8     8.2     8.3   8.0    8.7    8.3    NaN     8.8     8.3   
4    7.4    7.0     7.9     7.9   8.8    8.2    7.8    NaN     8.1     7.6   

   p_q11f  p_q1  p_q3d  p_q9  p_q10  p_q12aa  p_q12ab  p_q12ac  p_q12ad  \
0     7.7   NaN    NaN   NaN    NaN      NaN      NaN      NaN      NaN   
1     7.7   NaN    NaN   NaN    NaN      NaN      NaN      NaN      NaN   
2     8.6   NaN    NaN   NaN    NaN      NaN      NaN      NaN      NaN   
3     8.5   Na

# 9. Cleaning Up the Surveys

In [7]:
survey.shape

(1702, 2773)

There are two immediate facts that we can see in the data:

* There are over  `2000`  columns, nearly all of which we don't need. We'll have to filter the data to remove the unnecessary ones. Working with fewer columns will make it easier to print the dataframe out and find correlations within it.
* The survey data has a  `dbn`  column that we'll want to convert to uppercase ( `DBN` ). The conversion will make the column name consistent with the other data sets.

First, we'll need to filter the columns to remove the ones we don't need. Luckily, there's a data dictionary at the [original data download location](https://data.cityofnewyork.us/Education/NYC-School-Survey-2011/mnz3-dyi8). The dictionary tells us what each column represents. Based on our knowledge of the problem and the analysis we're trying to do, we can use the data dictionary to determine which columns to use

Here's a preview of the data dictionary:

![preview of the data dictionary](https://s3.amazonaws.com/dq-content/xj5ud4r.png "preview of the data dictionary")

Based on the dictionary, it looks like these are the relevant columns:
```python
["dbn", "rr_s", "rr_t", "rr_p", "N_s", "N_t", "N_p", "saf_p_11", "com_p_11", "eng_p_11", "aca_p_11", "saf_t_11", "com_t_11", "eng_t_11", "aca_t_11", "saf_s_11", "com_s_11", "eng_s_11", "aca_s_11", "saf_tot_11", "com_tot_11", "eng_tot_11", "aca_tot_11"]
```
These columns will give us aggregate survey data about how parents, teachers, and students feel about school safety, academic performance, and more. It will also give us the  `DBN` , which allows us to uniquely identify the school.

Before we filter columns out, we'll want to copy the data from the  `dbn`  column into a new column called  `DBN` .

In [8]:
survey["DBN"] = survey.dbn
survey = survey.loc[:, ["DBN", "rr_s", "rr_t", "rr_p", "N_s", "N_t", "N_p", "saf_p_11", "com_p_11", "eng_p_11", "aca_p_11", "saf_t_11", "com_t_11", "eng_t_11", "aca_t_11", "saf_s_11", "com_s_11", "eng_s_11", "aca_s_11", "saf_tot_11", "com_tot_11", "eng_tot_11", "aca_tot_11"]]
data['survey']=survey
data['survey'].shape

(1702, 23)

# 10. Inserting DBN fields

When we explored all of the data sets, we noticed that some of them, like  `class_size`  and  `hs_directory` , don't have a  `DBN`  column.  `hs_directory`  does have a  `dbn`  column, though, so we can just rename it.

In [9]:
data['hs_directory'].head(1)

dbn                          school_name      boro building_code  \
0  17K548  Brooklyn School for Music & Theatre  Brooklyn          K440   

   phone_number    fax_number grade_span_min  grade_span_max  \
0  718-230-6250  718-230-6262              9              12   

  expgrade_span_min  expgrade_span_max                                    bus  \
0               NaN                NaN  B41, B43, B44-SBS, B45, B48, B49, B69   

                                                       subway  \
0  2, 3, 4, 5, F, S to Botanic Garden ; B, Q to Prospect Park   

  primary_address_line_1      city state_code    zip              website  \
0     883 Classon Avenue  Brooklyn         NY  11225  Bkmusicntheatre.com   

   total_students                          campus_name school_type  \
0           399.0  Prospect Heights Educational Campus         NaN   

                                                                                                                                                                                                                                                                                                                                                                                                                      overview_paragraph  \
0  Brooklyn School for Music & Theatre (BSMT) uses our academic program to accommodate the intellectual, social, emotional and physical needs of creative high school students. Our vision is to provide a model professional environment where respect is mutual, ideas are shared and learning is not limited to the classroom. We prepare students for higher education and professional careers in the music and theatre industries.   

                                                                                                                                                                                                                                                   program_highlights  \
0  We offer highly competitive positions in our Drama, Chorus and Dance Company wherein students receive small group instruction focused on sharpening and furthering their skills while developing their professional portfolio for auditions in their chosen field.   

  language_classes                                advancedplacement_courses  \
0          Spanish  English Language and Composition, United States History   

  online_ap_courses online_language_courses  \
0               NaN                     NaN   

                                                                                                                                                                                                                                                                                                                                                                                                                                                                extracurricular_activities  \
0  Variety of clubs: Chess, The Step Team, Fashion, Tech Team, Women's Group; Extensive arts after-school program: Dance Company, Drama Company and Chorus Company, back stage crew program that trains students in running the lights, sound, video and all back stage and pit crew responsibilities; Saturday and After-school classes for Regents Preparation; School Leadership Team; Student Government; at least three annual major school-wide productions; two annual talent shows   

                                                                                       psal_sports_boys  \
0  Baseball, Basketball & JV Basketball, Cross Country, Indoor Track, Outdoor Track, Soccer, Volleyball   

                                                          psal_sports_girls  \
0  Basketball, Cross Country, Indoor Track, Outdoor Track, Soccer, Softball   

  psal_sports_coed school_sports  \
0              NaN           NaN   

                                                                                            

However,  `class_size`  doesn't appear to have the column at all. Here are the first few rows of the data set:

In [10]:
data['class_size'].head()

CSD BOROUGH SCHOOL CODE                SCHOOL NAME GRADE  PROGRAM TYPE  \
0    1       M        M015  P.S. 015 Roberto Clemente     0K       GEN ED   
1    1       M        M015  P.S. 015 Roberto Clemente     0K          CTT   
2    1       M        M015  P.S. 015 Roberto Clemente     01       GEN ED   
3    1       M        M015  P.S. 015 Roberto Clemente     01          CTT   
4    1       M        M015  P.S. 015 Roberto Clemente     02       GEN ED   

  CORE SUBJECT (MS CORE and 9-12 ONLY) CORE COURSE (MS CORE and 9-12 ONLY)  \
0                                    -                                   -   
1                                    -                                   -   
2                                    -                                   -   
3                                    -                                   -   
4                                    -                                   -   

  SERVICE CATEGORY(K-9* ONLY)  NUMBER OF STUDENTS / SEATS FILLED  \
0                           -                               19.0   
1                           -                               21.0   
2                           -                               17.0   
3                           -                               17.0   
4                           -                               15.0   

   NUMBER OF SECTIONS  AVERAGE CLASS SIZE  SIZE OF SMALLEST CLASS  \
0                 1.0                19.0                    19.0   
1                 1.0                21.0                    21.0   
2                 1.0                17.0                    17.0   
3                 1.0                17.0                    17.0   
4                 1.0                15.0                    15.0   

   SIZE OF LARGEST CLASS DATA SOURCE  SCHOOLWIDE PUPIL-TEACHER RATIO  
0                   19.0         ATS                             NaN  
1                   21.0         ATS                             NaN  
2                   17.0         ATS                             NaN  
3                   17.0         ATS                             NaN  
4                   15.0         ATS                             NaN

Here are the first few rows of the  `sat_results`  data, which does have a  `DBN`  column:

In [11]:
data['sat_results'].head()

DBN                                    SCHOOL NAME  \
0  01M292  HENRY STREET SCHOOL FOR INTERNATIONAL STUDIES   
1  01M448            UNIVERSITY NEIGHBORHOOD HIGH SCHOOL   
2  01M450                     EAST SIDE COMMUNITY SCHOOL   
3  01M458                      FORSYTH SATELLITE ACADEMY   
4  01M509                        MARTA VALLE HIGH SCHOOL   

  Num of SAT Test Takers SAT Critical Reading Avg. Score SAT Math Avg. Score  \
0                     29                             355                 404   
1                     91                             383                 423   
2                     70                             377                 402   
3                      7                             414                 401   
4                     44                             390                 433   

  SAT Writing Avg. Score  
0                    363  
1                    366  
2                    370  
3                    359  
4                    384

From looking at these rows, we can tell that the  `DBN`  in the  `sat_results`  data is just a combination of the  `CSD`  and  `SCHOOL CODE`  columns in the  `class_size`  data. The main difference is that the  `DBN`  is padded, so that the  `CSD`  portion of it always consists of two digits. That means we'll need to add a leading  `0`  to the  `CSD`  if the  `CSD`  is less than two digits long.

We can accomplish this using the [pandas.Series.apply()](https://pandas.pydata.org/pandas-docs/stable/generated/pandas.Series.apply.html) method, along with a custom function that:

* Takes in a number.
* Converts the number to a string using the [str()](https://docs.python.org/3/library/stdtypes.html#str) function.
* Check the length of the string using the [len()](https://docs.python.org/3/library/functions.html#len) function.
  * If the string is two digits long, returns the string.
  * If the string is one digit long, adds a  `0`  to the front of the string, then returns it.
    * we can use the string method [zfill()](https://docs.python.org/3/library/stdtypes.html#str.zfill) to do this.

Once we've padded the  `CSD` , we can use the addition operator ( `+` ) to combine the values in the  `CSD`  and  `SCHOOL CODE`  columns.

In [12]:
data['hs_directory']['DBN'] = data['hs_directory'].dbn 
data['class_size']['padded_csd'] = data['class_size'].CSD.apply(lambda x: str(x).zfill(2))
data['class_size']['DBN'] = data['class_size'].padded_csd + data['class_size']['SCHOOL CODE']
data['class_size'].head()

CSD BOROUGH SCHOOL CODE                SCHOOL NAME GRADE  PROGRAM TYPE  \
0    1       M        M015  P.S. 015 Roberto Clemente     0K       GEN ED   
1    1       M        M015  P.S. 015 Roberto Clemente     0K          CTT   
2    1       M        M015  P.S. 015 Roberto Clemente     01       GEN ED   
3    1       M        M015  P.S. 015 Roberto Clemente     01          CTT   
4    1       M        M015  P.S. 015 Roberto Clemente     02       GEN ED   

  CORE SUBJECT (MS CORE and 9-12 ONLY) CORE COURSE (MS CORE and 9-12 ONLY)  \
0                                    -                                   -   
1                                    -                                   -   
2                                    -                                   -   
3                                    -                                   -   
4                                    -                                   -   

  SERVICE CATEGORY(K-9* ONLY)  NUMBER OF STUDENTS / SEATS FILLED  \
0                           -                               19.0   
1                           -                               21.0   
2                           -                               17.0   
3                           -                               17.0   
4                           -                               15.0   

   NUMBER OF SECTIONS  AVERAGE CLASS SIZE  SIZE OF SMALLEST CLASS  \
0                 1.0                19.0                    19.0   
1                 1.0                21.0                    21.0   
2                 1.0                17.0                    17.0   
3                 1.0                17.0                    17.0   
4                 1.0                15.0                    15.0   

   SIZE OF LARGEST CLASS DATA SOURCE  SCHOOLWIDE PUPIL-TEACHER RATIO  \
0                   19.0         ATS                             NaN   
1                   21.0         ATS                             NaN   
2                   17.0         ATS                             NaN   
3                   17.0         ATS                             NaN   
4                   15.0         ATS                             NaN   

  padded_csd     DBN  
0         01  01M015  
1         01  01M015  
2         01  01M015  
3         01  01M015  
4         01  01M015

# 12. Combining the SAT scores

Now we're almost ready to combine our data sets. Before we do, let's take some time to calculate variables that will be useful in our analysis. We've already discussed one such variable -- a column that totals up the SAT scores for the different sections of the exam. This will make it much easier to correlate scores with demographic factors because we'll be working with a single number, rather than three different ones.

Before we can generate this column, we'll need to convert the  `SAT Math Avg. Score` ,  `SAT Critical Reading Avg. Score` , and  `SAT Writing Avg. Score`  columns in the  `sat_results`  data set from the object (string) data type to a numeric data type. We can use the [pandas.to_numeric()](http://pandas.pydata.org/pandas-docs/version/0.17.0/generated/pandas.to_numeric.html) method for the conversion. If we don't convert the values, we won't be able to add the columns together.

It's important to pass the keyword argument  `errors="coerce"`  when we call  `pandas.to_numeric()` , so that pandas treats any invalid strings it can't convert to numbers as missing values instead.

After we perform the conversion, we can use the addition operator ( `+` ) to add all three columns together.

In [13]:
data['sat_results'][
    ['SAT Math Avg. Score', 'SAT Critical Reading Avg. Score', 'SAT Writing Avg. Score']
] = data['sat_results'][
    ['SAT Math Avg. Score', 'SAT Critical Reading Avg. Score', 'SAT Writing Avg. Score']
].apply(lambda x: pd.to_numeric(x, errors="coerce"))

data['sat_results']['sat_score'] = data['sat_results'][
    ['SAT Math Avg. Score', 'SAT Critical Reading Avg. Score', 'SAT Writing Avg. Score']
].sum(axis=1)

data['sat_results'].sat_score.head()

0    1122.0
1    1172.0
2    1149.0
3    1174.0
4    1207.0
Name: sat_score, dtype: float64

# 13. Parsing Geographic Coordinates for schools

Next, we'll want to parse the latitude and longitude coordinates for each school. This will enable us to map the schools and uncover any geographic patterns in the data. The coordinates are currently in the text field  `Location 1`  in the  `hs_directory`  data set.

Let's take a look at the first few rows:

In [14]:
print(data['hs_directory']['Location 1'].head())

0        883 Classon Avenue\nBrooklyn, NY 11225\n(40.67029890700047, -73.96164787599963)
1              1110 Boston Road\nBronx, NY 10456\n(40.8276026690005, -73.90447525699966)
2          1501 Jerome Avenue\nBronx, NY 10452\n(40.842414068000494, -73.91616158599965)
3          411 Pearl Street\nNew York, NY 10038\n(40.71067947100045, -74.00080702099967)
4    160-20 Goethals Avenue\nJamaica, NY 11432\n(40.718810094000446, -73.80650045499965)
Name: Location 1, dtype: object


As you can see, this field contains a lot of information we don't need. We want to extract the coordinates, which are in parentheses at the end of the field.

We want to extract the latitude, 40.8276026690005, and the longitude, -73.90447525699966. Taken together, latitude and longitude make up a pair of coordinates that allows us to pinpoint any location on Earth.

We can do the extraction with a regular expression.

In [15]:
data['hs_directory']['lat'] = data['hs_directory']['Location 1'].apply(
    lambda x: re.findall("\d+.\d+.,", (re.findall("\(.+\)", x))[0])[0].strip(","))
data['hs_directory'].lat.head()

0     40.67029890700047
1      40.8276026690005
2    40.842414068000494
3     40.71067947100045
4    40.718810094000446
Name: lat, dtype: object

# 14. Extracting the longitude

On the last screen, we parsed the latitude from the  `Location 1`  column. Now we'll just need to do the same for the longitude.

Once we have both coordinates, we'll need to convert them to numeric values. We can use the  `pandas.to_numeric()`  function to convert them from strings to numbers.

In [16]:
data['hs_directory']['lon'] = data['hs_directory']['Location 1'].apply(
    lambda x: re.findall("\(.+\)", x)[0].split(",")[-1].replace(")", "").strip())
data['hs_directory'].lon.head()
data['hs_directory'][['lat', 'lon']] = data['hs_directory'][['lat', 'lon']].apply(lambda x: pd.to_numeric(x, errors="coerce"))
data['hs_directory'].head(1)

dbn                          school_name      boro building_code  \
0  17K548  Brooklyn School for Music & Theatre  Brooklyn          K440   

   phone_number    fax_number grade_span_min  grade_span_max  \
0  718-230-6250  718-230-6262              9              12   

  expgrade_span_min  expgrade_span_max                                    bus  \
0               NaN                NaN  B41, B43, B44-SBS, B45, B48, B49, B69   

                                                       subway  \
0  2, 3, 4, 5, F, S to Botanic Garden ; B, Q to Prospect Park   

  primary_address_line_1      city state_code    zip              website  \
0     883 Classon Avenue  Brooklyn         NY  11225  Bkmusicntheatre.com   

   total_students                          campus_name school_type  \
0           399.0  Prospect Heights Educational Campus         NaN   

                                                                                                                                                                                                                                                                                                                                                                                                                      overview_paragraph  \
0  Brooklyn School for Music & Theatre (BSMT) uses our academic program to accommodate the intellectual, social, emotional and physical needs of creative high school students. Our vision is to provide a model professional environment where respect is mutual, ideas are shared and learning is not limited to the classroom. We prepare students for higher education and professional careers in the music and theatre industries.   

                                                                                                                                                                                                                                                   program_highlights  \
0  We offer highly competitive positions in our Drama, Chorus and Dance Company wherein students receive small group instruction focused on sharpening and furthering their skills while developing their professional portfolio for auditions in their chosen field.   

  language_classes                                advancedplacement_courses  \
0          Spanish  English Language and Composition, United States History   

  online_ap_courses online_language_courses  \
0               NaN                     NaN   

                                                                                                                                                                                                                                                                                                                                                                                                                                                                extracurricular_activities  \
0  Variety of clubs: Chess, The Step Team, Fashion, Tech Team, Women's Group; Extensive arts after-school program: Dance Company, Drama Company and Chorus Company, back stage crew program that trains students in running the lights, sound, video and all back stage and pit crew responsibilities; Saturday and After-school classes for Regents Preparation; School Leadership Team; Student Government; at least three annual major school-wide productions; two annual talent shows   

                                                                                       psal_sports_boys  \
0  Baseball, Basketball & JV Basketball, Cross Country, Indoor Track, Outdoor Track, Soccer, Volleyball   

                                                          psal_sports_girls  \
0  Basketball, Cross Country, Indoor Track, Outdoor Track, Soccer, Softball   

  psal_sports_coed school_sports  \
0              NaN           NaN   

                                                                                            

We're almost ready to combine our data sets! We've come a long way in this mission -- we've gone from choosing a topic for a project to acquiring the data to having clean data that we're almost ready to combine.

Along the way, we've explored how to:

* Handle files with different formats and columns
* Prepare to merge multiple files
* Use text processing to extract coordinates from a string
* Convert columns from strings to numbers


We'll finish cleaning the data sets, then combine them so we can start our analysis.

## Combining the Data

## Introductions

we began investigating possible relationships between SAT scores and demographic factors. In order to do this, we acquired several data sets about [New York City public schools](https://data.cityofnewyork.us/data?cat=education). We manipulated these data sets, and found that we could combine them all using the  `DBN`  column. All of the data sets are currently stored as keys in the  `data`  dictionary. Each individual data set is a pandas dataframe.

we'll clean the data a bit more, then combine it. Finally, we'll compute correlations and perform some analysis.

The first thing we'll need to do in preparation for the merge is condense some of the data sets. We noticed that the values in the  `DBN`  column were unique in the  `sat_results`  data set. Other data sets like  `class_size`  had duplicate  `DBN`  values, however.

We'll need to condense these data sets so that each value in the  `DBN`  column is unique. If not, we'll run into issues when it comes time to combine the data sets.

While the main data set we want to analyze,  `sat_results` , has unique  `DBN`  values for every high school in New York City, other data sets aren't as clean. A single row in the  `sat_results`  data set may match multiple rows in the  `class_size`  data set, for example. This situation will create problems, because we don't know which of the multiple entries in the  `class_size`  data set we should combine with the single matching entry in  `sat_results` .

In [17]:
data['sat_results'].DBN.head()

0    01M292
1    01M448
2    01M450
3    01M458
4    01M509
Name: DBN, dtype: object

In [18]:
data['class_size'].DBN.head()

0    01M015
1    01M015
2    01M015
3    01M015
4    01M015
Name: DBN, dtype: object

we can't just combine the rows from both data sets because there are several cases where multiple rows in  `class_size`  match a single row in  `sat_results` .

To resolve this issue, we'll condense the  `class_size` ,  `graduation` , and  `demographics`  data sets so that each  `DBN`  is unique.

## 2. Condensing the Classize datasets

The first data set that we'll condense is  `class_size` . The first few rows of  `class_size`  look like this:

In [19]:
data['class_size'].head()

CSD BOROUGH SCHOOL CODE                SCHOOL NAME GRADE  PROGRAM TYPE  \
0    1       M        M015  P.S. 015 Roberto Clemente     0K       GEN ED   
1    1       M        M015  P.S. 015 Roberto Clemente     0K          CTT   
2    1       M        M015  P.S. 015 Roberto Clemente     01       GEN ED   
3    1       M        M015  P.S. 015 Roberto Clemente     01          CTT   
4    1       M        M015  P.S. 015 Roberto Clemente     02       GEN ED   

  CORE SUBJECT (MS CORE and 9-12 ONLY) CORE COURSE (MS CORE and 9-12 ONLY)  \
0                                    -                                   -   
1                                    -                                   -   
2                                    -                                   -   
3                                    -                                   -   
4                                    -                                   -   

  SERVICE CATEGORY(K-9* ONLY)  NUMBER OF STUDENTS / SEATS FILLED  \
0                           -                               19.0   
1                           -                               21.0   
2                           -                               17.0   
3                           -                               17.0   
4                           -                               15.0   

   NUMBER OF SECTIONS  AVERAGE CLASS SIZE  SIZE OF SMALLEST CLASS  \
0                 1.0                19.0                    19.0   
1                 1.0                21.0                    21.0   
2                 1.0                17.0                    17.0   
3                 1.0                17.0                    17.0   
4                 1.0                15.0                    15.0   

   SIZE OF LARGEST CLASS DATA SOURCE  SCHOOLWIDE PUPIL-TEACHER RATIO  \
0                   19.0         ATS                             NaN   
1                   21.0         ATS                             NaN   
2                   17.0         ATS                             NaN   
3                   17.0         ATS                             NaN   
4                   15.0         ATS                             NaN   

  padded_csd     DBN  
0         01  01M015  
1         01  01M015  
2         01  01M015  
3         01  01M015  
4         01  01M015

As you can see, the first few rows all pertain to the same school, which is why the  `DBN`  appears more than once. It looks like each school has multiple values for  `GRADE` ,  `PROGRAM TYPE` ,  `CORE SUBJECT (MS CORE and 9-12 ONLY)` , and  `CORE COURSE (MS CORE and 9-12 ONLY)` .

If we look at the unique values for  `GRADE` , we get the following:

In [20]:
data['class_size']['GRADE '].unique()

array(['0K', '01', '02', '03', '04', '05', '0K-09', nan, '06', '07', '08',
       'MS Core', '09-12', '09'], dtype=object)

Because we're dealing with high schools, we're only concerned with grades  `9`  through  `12` . That means we only want to pick rows where the value in the  `GRADE`  column is  `09-12`

If we look at the unique values for  `PROGRAM TYPE` , we get the following:

In [21]:
data['class_size']['PROGRAM TYPE'].unique()

array(['GEN ED', 'CTT', 'SPEC ED', nan, 'G&T'], dtype=object)

In [22]:
data['class_size']['PROGRAM TYPE'].value_counts()

GEN ED     14545
CTT         7460
SPEC ED     3653
G&T          469
Name: PROGRAM TYPE, dtype: int64

Each school can have multiple program types. Because  `GEN ED`  is the largest category by far, let's only select rows where  `PROGRAM TYPE`  is  `GEN ED` .

In [23]:
data['class_size'].shape

(27611, 18)

In [24]:
class_size = data['class_size']
class_size = class_size[(class_size['GRADE ']=='09-12') & (class_size['PROGRAM TYPE']=='GEN ED')]
class_size.head()

CSD BOROUGH SCHOOL CODE                                    SCHOOL NAME  \
225    1       M        M292  Henry Street School for International Studies   
226    1       M        M292  Henry Street School for International Studies   
227    1       M        M292  Henry Street School for International Studies   
228    1       M        M292  Henry Street School for International Studies   
229    1       M        M292  Henry Street School for International Studies   

    GRADE  PROGRAM TYPE CORE SUBJECT (MS CORE and 9-12 ONLY)  \
225  09-12       GEN ED                              ENGLISH   
226  09-12       GEN ED                              ENGLISH   
227  09-12       GEN ED                              ENGLISH   
228  09-12       GEN ED                              ENGLISH   
229  09-12       GEN ED                                 MATH   

    CORE COURSE (MS CORE and 9-12 ONLY) SERVICE CATEGORY(K-9* ONLY)  \
225                           English 9                           -   
226                          English 10                           -   
227                          English 11                           -   
228                          English 12                           -   
229                  Integrated Algebra                           -   

     NUMBER OF STUDENTS / SEATS FILLED  NUMBER OF SECTIONS  \
225                               63.0                 3.0   
226                               79.0                 3.0   
227                               38.0                 2.0   
228                               69.0                 3.0   
229                               53.0                 3.0   

     AVERAGE CLASS SIZE  SIZE OF SMALLEST CLASS  SIZE OF LARGEST CLASS  \
225                21.0                    19.0                   25.0   
226                26.3                    24.0                   31.0   
227                19.0                    16.0                   22.0   
228                23.0                    13.0                   30.0   
229                17.7                    16.0                   21.0   

    DATA SOURCE  SCHOOLWIDE PUPIL-TEACHER RATIO padded_csd     DBN  
225       STARS                             NaN         01  01M292  
226       STARS                             NaN         01  01M292  
227       STARS                             NaN         01  01M292  
228       STARS                             NaN         01  01M292  
229       STARS                             NaN         01  01M292

In [25]:
class_size.shape

(6513, 18)

## 3. Computing Average Class Sizes

As we saw when we displayed  `class_size`  on the last screen,  `DBN`  still isn't completely unique. This is due to the  `CORE COURSE (MS CORE and 9-12 ONLY)`  and  `CORE SUBJECT (MS CORE and 9-12 ONLY)`  columns.

`CORE COURSE (MS CORE and 9-12 ONLY)`  and  `CORE SUBJECT (MS CORE and 9-12 ONLY)`  seem to pertain to different kinds of classes. For example, here are the unique values for  `CORE SUBJECT (MS CORE and 9-12 ONLY)` :

In [26]:
class_size['CORE SUBJECT (MS CORE and 9-12 ONLY)'].unique()

array(['ENGLISH', 'MATH', 'SCIENCE', 'SOCIAL STUDIES'], dtype=object)

This column only seems to include certain subjects. We want our class size data to include every single class a school offers -- not just a subset of them. What we can do is take the average across all of the classes a school offers. This will give us unique  `DBN`  values, while also incorporating as much data as possible into the average.

Fortunately, we can use the [pandas.DataFrame.groupby()](http://pandas.pydata.org/pandas-docs/stable/groupby.html) method to help us with this. The  `DataFrame.groupby()`  method will split a dataframe up into unique groups, based on a given column. We can then use the [agg()](http://pandas.pydata.org/pandas-docs/stable/groupby.html#aggregation) method on the resulting  `pandas.core.groupby`  object to find the mean of each column.

In [27]:
class_size = class_size.groupby('DBN').agg(np.mean).reset_index()
data['class_size'] = class_size
data['class_size'].head()

DBN  CSD  NUMBER OF STUDENTS / SEATS FILLED  NUMBER OF SECTIONS  \
0  01M292    1                            88.0000            4.000000   
1  01M332    1                            46.0000            2.000000   
2  01M378    1                            33.0000            1.000000   
3  01M448    1                           105.6875            4.750000   
4  01M450    1                            57.6000            2.733333   

   AVERAGE CLASS SIZE  SIZE OF SMALLEST CLASS  SIZE OF LARGEST CLASS  \
0           22.564286                   18.50              26.571429   
1           22.000000                   21.00              23.500000   
2           33.000000                   33.00              33.000000   
3           22.231250                   18.25              27.062500   
4           21.200000                   19.40              22.866667   

   SCHOOLWIDE PUPIL-TEACHER RATIO  
0                             NaN  
1                             NaN  
2                             NaN  
3                             NaN  
4                             NaN

## 4. Condensing Demographics Dataset

Now that we've finished condensing  `class_size` , let's condense  `demographics` . The first few rows look like this:

In [28]:
data['demographics'].head()

DBN                       Name  schoolyear fl_percent  frl_percent  \
0  01M015  P.S. 015 ROBERTO CLEMENTE    20052006       89.4          NaN   
1  01M015  P.S. 015 ROBERTO CLEMENTE    20062007       89.4          NaN   
2  01M015  P.S. 015 ROBERTO CLEMENTE    20072008       89.4          NaN   
3  01M015  P.S. 015 ROBERTO CLEMENTE    20082009       89.4          NaN   
4  01M015  P.S. 015 ROBERTO CLEMENTE    20092010                    96.5   

   total_enrollment prek   k grade1 grade2 grade3 grade4 grade5 grade6 grade7  \
0               281   15  36     40     33     38     52     29     38    NaN   
1               243   15  29     39     38     34     42     46    NaN    NaN   
2               261   18  43     39     36     38     47     40    NaN    NaN   
3               252   17  37     44     32     34     39     49    NaN    NaN   
4               208   16  40     28     32     30     24     38    NaN    NaN   

  grade8 grade9 grade10 grade11 grade12  ell_num  ell_percent  sped_num  \
0    NaN    NaN     NaN     NaN     NaN     36.0         12.8      57.0   
1    NaN    NaN     NaN     NaN     NaN     38.0         15.6      55.0   
2    NaN    NaN     NaN     NaN     NaN     52.0         19.9      60.0   
3    NaN    NaN     NaN     NaN     NaN     48.0         19.0      62.0   
4    NaN    NaN     NaN     NaN     NaN     40.0         19.2      46.0   

   sped_percent ctt_num selfcontained_num  asian_num  asian_per  black_num  \
0          20.3      25                 9         10        3.6         74   
1          22.6      19                15         18        7.4         68   
2          23.0      20                14         16        6.1         77   
3          24.6      21                17         16        6.3         75   
4          22.1      14                14         16        7.7         67   

   black_per  hispanic_num  hispanic_per  white_num  white_per  male_num  \
0       26.3           189          67.3          5        1.8     158.0   
1       28.0           153          63.0          4        1.6     140.0   
2       29.5           157          60.2          7        2.7     143.0   
3       29.8           149          59.1          7        2.8     149.0   
4       32.2           118          56.7          6        2.9     124.0   

   male_per  female_num  female_per  
0      56.2       123.0        43.8  
1      57.6       103.0        42.4  
2      54.8       118.0        45.2  
3      59.1       103.0        40.9  
4      59.6        84.0        40.4

In [29]:
data['demographics'].shape

(10075, 38)

In this case, the only column that prevents a given  `DBN`  from being unique is  `schoolyear` . We only want to select rows where  `schoolyear`  is  `20112012` . This will give us the most recent year of data, and also match our SAT results data.

In [30]:
data['demographics'] = data['demographics'][data['demographics'].schoolyear==20112012]
data['demographics'].head()

DBN                                              Name  schoolyear  \
6   01M015  P.S. 015 ROBERTO CLEMENTE                           20112012   
13  01M019  P.S. 019 ASHER LEVY                                 20112012   
20  01M020  PS 020 ANNA SILVER                                  20112012   
27  01M034  PS 034 FRANKLIN D ROOSEVELT                         20112012   
35  01M063  PS 063 WILLIAM MCKINLEY                             20112012   

   fl_percent  frl_percent  total_enrollment prek    k grade1 grade2 grade3  \
6         NaN         89.4               189   13   31     35     28     25   
13        NaN         61.5               328   32   46     52     54     52   
20        NaN         92.5               626   52  102    121     87     88   
27        NaN         99.7               401   14   34     38     36     45   
35        NaN         78.9               176   18   20     30     21     31   

   grade4 grade5 grade6 grade7 grade8 grade9 grade10 grade11 grade12  ell_num  \
6      28     29                                                         20.0   
13     46     46                                                         33.0   
20     85     91                                                        128.0   
27     28     40     55     55     56                                    34.0   
35     26     30                                                          6.0   

    ell_percent  sped_num  sped_percent ctt_num selfcontained_num  asian_num  \
6          10.6      40.0          21.2      23                 7         12   
13         10.1      59.0          18.0      16                16         51   
20         20.4      97.0          15.5      49                31        190   
27          8.5     106.0          26.4      59                16         22   
35          3.4      45.0          25.6      34                 4          9   

    asian_per  black_num  black_per  hispanic_num  hispanic_per  white_num  \
6         6.3         63       33.3           109          57.7          4   
13       15.5         81       24.7           158          48.2         28   
20       30.4         55        8.8           357          57.0         16   
27        5.5         90       22.4           275          68.6          8   
35        5.1         41       23.3           110          62.5         15   

    white_per  male_num  male_per  female_num  female_per  
6         2.1      97.0      51.3        92.0        48.7  
13        8.5     147.0      44.8       181.0        55.2  
20        2.6     330.0      52.7       296.0        47.3  
27        2.0     204.0      50.9       197.0        49.1  
35        8.5      97.0      55.1        79.0        44.9

## 5. Condensing graduation dataset

Finally, we'll need to condense the  `graduation`  data set. Here are the first few rows:

In [31]:
data['graduation'].head()

Demographic     DBN                            School Name    Cohort  \
0  Total Cohort  01M292  HENRY STREET SCHOOL FOR INTERNATIONAL      2003   
1  Total Cohort  01M292  HENRY STREET SCHOOL FOR INTERNATIONAL      2004   
2  Total Cohort  01M292  HENRY STREET SCHOOL FOR INTERNATIONAL      2005   
3  Total Cohort  01M292  HENRY STREET SCHOOL FOR INTERNATIONAL      2006   
4  Total Cohort  01M292  HENRY STREET SCHOOL FOR INTERNATIONAL  2006 Aug   

   Total Cohort Total Grads - n Total Grads - % of cohort Total Regents - n  \
0             5               s                         s                 s   
1            55              37                     67.3%                17   
2            64              43                     67.2%                27   
3            78              43                     55.1%                36   
4            78              44                     56.4%                37   

  Total Regents - % of cohort Total Regents - % of grads Advanced Regents - n  \
0                           s                          s                    s   
1                       30.9%                      45.9%                    0   
2                       42.2%                      62.8%                    0   
3                       46.2%                      83.7%                    0   
4                       47.4%                      84.1%                    0   

  Advanced Regents - % of cohort Advanced Regents - % of grads  \
0                              s                             s   
1                             0%                            0%   
2                             0%                            0%   
3                             0%                            0%   
4                             0%                            0%   

  Regents w/o Advanced - n Regents w/o Advanced - % of cohort  \
0                        s                                  s   
1                       17                              30.9%   
2                       27                              42.2%   
3                       36                              46.2%   
4                       37                              47.4%   

  Regents w/o Advanced - % of grads Local - n Local - % of cohort  \
0                                 s         s                   s   
1                             45.9%        20               36.4%   
2                             62.8%        16                 25%   
3                             83.7%         7                  9%   
4                             84.1%         7                  9%   

    Local - % of grads Still Enrolled - n Still Enrolled - % of cohort  \
0                    s                  s                            s   
1                54.1%                 15                        27.3%   
2  37.200000000000003%                  9                        14.1%   
3                16.3%                 16                        20.5%   
4                15.9%                 15                        19.2%   

  Dropped Out - n Dropped Out - % of cohort  
0               s                         s  
1               3                      5.5%  
2               9                     14.1%  
3              11                     14.1%  
4              11                     14.1%

The  `Demographic`  and  `Cohort`  columns are what prevent  `DBN`  from being unique in the  `graduation`  data. A  `Cohort`  appears to refer to the year the data represents, and the  `Demographic`  appears to refer to a specific demographic group. In this case, we want to pick data from the most recent  `Cohort`  available, which is  `2006` . We also want data from the full cohort, so we'll only pick rows where  `Demographic`  is  `Total Cohort` .

In [32]:
data['graduation'] = data['graduation'][(data['graduation'].Cohort=='2006') & (data['graduation'].Demographic=='Total Cohort')]
data['graduation'].head()

Demographic     DBN                            School Name Cohort  \
3   Total Cohort  01M292  HENRY STREET SCHOOL FOR INTERNATIONAL   2006   
10  Total Cohort  01M448    UNIVERSITY NEIGHBORHOOD HIGH SCHOOL   2006   
17  Total Cohort  01M450             EAST SIDE COMMUNITY SCHOOL   2006   
24  Total Cohort  01M509                MARTA VALLE HIGH SCHOOL   2006   
31  Total Cohort  01M515  LOWER EAST SIDE PREPARATORY HIGH SCHO   2006   

    Total Cohort Total Grads - n Total Grads - % of cohort Total Regents - n  \
3             78              43                     55.1%                36   
10           124              53                     42.7%                42   
17            90              70                     77.8%                67   
24            84              47                       56%                40   
31           193             105                     54.4%                91   

   Total Regents - % of cohort Total Regents - % of grads  \
3                        46.2%                      83.7%   
10                       33.9%                      79.2%   
17         74.400000000000006%                      95.7%   
24                       47.6%                      85.1%   
31                       47.2%                      86.7%   

   Advanced Regents - n Advanced Regents - % of cohort  \
3                     0                             0%   
10                    8                           6.5%   
17                    0                             0%   
24                   17                          20.2%   
31                   69            35.799999999999997%   

   Advanced Regents - % of grads Regents w/o Advanced - n  \
3                             0%                       36   
10                         15.1%                       34   
17                            0%                       67   
24           36.200000000000003%                       23   
31                         65.7%                       22   

   Regents w/o Advanced - % of cohort Regents w/o Advanced - % of grads  \
3                               46.2%                             83.7%   
10                              27.4%                             64.2%   
17                74.400000000000006%                             95.7%   
24                              27.4%                             48.9%   
31                              11.4%                               21%   

   Local - n Local - % of cohort Local - % of grads Still Enrolled - n  \
3          7                  9%              16.3%                 16   
10        11                8.9%              20.8%                 46   
17         3                3.3%               4.3%                 15   
24         7  8.300000000000001%              14.9%                 25   
31        14                7.3%              13.3%                 53   

   Still Enrolled - % of cohort Dropped Out - n Dropped Out - % of cohort  
3                         20.5%              11                     14.1%  
10                        37.1%              20       16.100000000000001%  
17                        16.7%               5                      5.6%  
24                        29.8%               5                        6%  
31                        27.5%              35       18.100000000000001%

# 6. Converting AP Scores

We're almost ready to combine all of the data sets. The only remaining thing to do is convert the [Advanced Placement](https://en.wikipedia.org/wiki/Advanced_Placement_exams) (AP) test scores from strings to numeric values. High school students take the AP exams before applying to college. There are several AP exams, each corresponding to a school subject. High school students who earn high scores may receive college credit.

AP exams have a  `1`  to  `5`  scale;  `3`  or higher is a passing score. Many high school students take AP exams -- particularly those who attend academically challenging institutions. AP exams are much more rare in schools that lack funding or academic rigor.

It will be interesting to find out whether AP exam scores are correlated with SAT scores across high schools. To determine this, we'll need to convert the AP exam scores in the  `ap_2010`  data set to numeric values first.

There are three columns we'll need to convert:

* `AP Test Takers`
* `Total Exams Taken`
* `Number of Exams with scores 3 4 or 5`

In [33]:
data['ap_2010'].head()

DBN                             SchoolName AP Test Takers   \
0  01M448           UNIVERSITY NEIGHBORHOOD H.S.              39   
1  01M450                 EAST SIDE COMMUNITY HS              19   
2  01M515                    LOWER EASTSIDE PREP              24   
3  01M539         NEW EXPLORATIONS SCI,TECH,MATH             255   
4  02M296  High School of Hospitality Management               s   

  Total Exams Taken Number of Exams with scores 3 4 or 5  
0                49                                   10  
1                21                                    s  
2                26                                   24  
3               377                                  191  
4                 s                                    s

In [34]:
cols = ['AP Test Takers ', 'Total Exams Taken', 'Number of Exams with scores 3 4 or 5']
data['ap_2010'][cols] = data['ap_2010'][cols].apply(lambda x: pd.to_numeric(x, errors='coerce'))
data['ap_2010'].head()

DBN                             SchoolName  AP Test Takers   \
0  01M448           UNIVERSITY NEIGHBORHOOD H.S.             39.0   
1  01M450                 EAST SIDE COMMUNITY HS             19.0   
2  01M515                    LOWER EASTSIDE PREP             24.0   
3  01M539         NEW EXPLORATIONS SCI,TECH,MATH            255.0   
4  02M296  High School of Hospitality Management              NaN   

   Total Exams Taken  Number of Exams with scores 3 4 or 5  
0               49.0                                  10.0  
1               21.0                                   NaN  
2               26.0                                  24.0  
3              377.0                                 191.0  
4                NaN                                   NaN

## 7. Left, Right, Inner & outer joins

Before we merge our data, we'll need to decide on the merge strategy we want to use. We'll be using the pandas [pandas.DataFrame.merge()](http://pandas.pydata.org/pandas-docs/stable/generated/pandas.DataFrame.merge.html) function, which supports four types of joins --  `left` ,  `right` ,  `inner` , and  `outer` . Each of these join types dictates how pandas combines the rows.

We'll be using the  `DBN`  column to identify matching rows across data sets. In other words, the values in that column will help us know which row from the first data set to combine with which row in the second data set.

There may be  `DBN`  values that exist in one data set but not in another. This is partly because the data is from different years. Each data set also has inconsistencies in terms of how it was gathered. Human error (and other types of errors) may also play a role. Therefore, we may not find matches for the  `DBN`  values in  `sat_results`  in all of the other data sets, and other data sets may have  `DBN`  values that don't exist in  `sat_results` .

We'll merge two data sets at a time. For example, we'll merge  `sat_results`  and  `hs_directory` , then merge the result with  `ap_2010` , then merge the result of that with  `class_size` . We'll continue combining data sets in this way until we've merged all of them. Afterwards, we'll have roughly the same number of rows, but each row will have columns from all of the data sets.

Because this project is concerned with determing demographic factors that correlate with SAT score, we'll want to preserve as many rows as possible from  `sat_results`  while minimizing null values.

This means that we may need to use different merge strategies with different data sets. Some of the data sets have a lot of missing  `DBN`  values. This makes a  `left`  join more appropriate, because we don't want to lose too many rows when we merge. If we did an  `inner`  join, we would lose the data for many high schools.

Some data sets have  `DBN`  values that are almost identical to those in  `sat_results` . Those data sets also have information we need to keep. Most of our analysis would be impossible if a significant number of rows was missing from  `demographics` , for example. Therefore, we'll do an inner join to avoid missing data in these columns.

# 8. Performing Left joins

Both the  `ap_2010`  and the  `graduation`  data sets have many missing  `DBN`  values, so we'll use a  `left`  join when we merge the  `sat_results`  data set with them. Because we're using a  `left`  join, our final dataframe will have all of the same  `DBN`  values as the original  `sat_results`  dataframe.

We'll need to use the pandas  `df.merge()`  method to merge dataframes. The "left" dataframe is the one we call the method on, and the "right" dataframe is the one we pass into  `df.merge()` .

Because we're using the  `DBN`  column to join the dataframes, we'll need to specify the keyword argument  `on="DBN"`  when calling  `pandas.DataFrame.merge()` .

First, we'll assign  `data["sat_results"]`  to the variable  `combined` . Then, we'll merge all of the other dataframes with  `combined` . When we're finished,  `combined`  will have all of the columns from all of the data sets.

In [35]:
combined = data["sat_results"].merge(data['ap_2010'], how='left', on='DBN')
combined = combined.merge(data['graduation'], how='left', on='DBN')
combined.head()

DBN                                    SCHOOL NAME  \
0  01M292  HENRY STREET SCHOOL FOR INTERNATIONAL STUDIES   
1  01M448            UNIVERSITY NEIGHBORHOOD HIGH SCHOOL   
2  01M450                     EAST SIDE COMMUNITY SCHOOL   
3  01M458                      FORSYTH SATELLITE ACADEMY   
4  01M509                        MARTA VALLE HIGH SCHOOL   

  Num of SAT Test Takers  SAT Critical Reading Avg. Score  \
0                     29                            355.0   
1                     91                            383.0   
2                     70                            377.0   
3                      7                            414.0   
4                     44                            390.0   

   SAT Math Avg. Score  SAT Writing Avg. Score  sat_score  \
0                404.0                   363.0     1122.0   
1                423.0                   366.0     1172.0   
2                402.0                   370.0     1149.0   
3                401.0                   359.0     1174.0   
4                433.0                   384.0     1207.0   

                     SchoolName  AP Test Takers   Total Exams Taken  \
0                           NaN              NaN                NaN   
1  UNIVERSITY NEIGHBORHOOD H.S.             39.0               49.0   
2        EAST SIDE COMMUNITY HS             19.0               21.0   
3                           NaN              NaN                NaN   
4                           NaN              NaN                NaN   

   Number of Exams with scores 3 4 or 5   Demographic  \
0                                   NaN  Total Cohort   
1                                  10.0  Total Cohort   
2                                   NaN  Total Cohort   
3                                   NaN           NaN   
4                                   NaN  Total Cohort   

                             School Name Cohort  Total Cohort Total Grads - n  \
0  HENRY STREET SCHOOL FOR INTERNATIONAL   2006          78.0              43   
1    UNIVERSITY NEIGHBORHOOD HIGH SCHOOL   2006         124.0              53   
2             EAST SIDE COMMUNITY SCHOOL   2006          90.0              70   
3                                    NaN    NaN           NaN             NaN   
4                MARTA VALLE HIGH SCHOOL   2006          84.0              47   

  Total Grads - % of cohort Total Regents - n Total Regents - % of cohort  \
0                     55.1%                36                       46.2%   
1                     42.7%                42                       33.9%   
2                     77.8%                67         74.400000000000006%   
3                       NaN               NaN                         NaN   
4                       56%                40                       47.6%   

  Total Regents - % of grads Advanced Regents - n  \
0                      83.7%                    0   
1                      79.2%                    8   
2                      95.7%                    0   
3                        NaN                  NaN   
4                      85.1%                   17   

  Advanced Regents - % of cohort Advanced Regents - % of grads  \
0                             0%                            0%   
1                           6.5%                         15.1%   
2                             0%                            0%   
3                            NaN                           NaN   
4                          20.2%           36.200000000000003%   

  Regents w/o Advanced - n Regents w/o Advanced - % of cohort  \
0                       36                              46.2%   
1                       34                              27.4%   
2                       67                74.400000000000006%   
3                      NaN                                NaN   
4                       23                              27.4%   

  Regents w/o Advanced - % of grads Local - n Local - % of cohort  \
0              

In [36]:
combined.shape

(479, 33)

## 9. Performing Inner Joins

Now that we've performed the  `left`  joins, we still have to merge  `class_size` ,  `demographics` ,  `survey` , and  `hs_directory`  into  `combined` . Because these files contain information that's more valuable to our analysis and also have fewer missing  `DBN`  values, we'll use the  `inner`  join type.

In [37]:
combined = combined.merge(data['class_size'], how='inner', on='DBN')
combined = combined.merge(data['demographics'], how='inner', on='DBN')
combined = combined.merge(data['survey'], how='inner', on='DBN')
combined = combined.merge(data['hs_directory'], how='inner', on='DBN')
combined.head()

DBN                                                     SCHOOL NAME  \
0  01M292                   HENRY STREET SCHOOL FOR INTERNATIONAL STUDIES   
1  01M448                             UNIVERSITY NEIGHBORHOOD HIGH SCHOOL   
2  01M450                                      EAST SIDE COMMUNITY SCHOOL   
3  01M509                                         MARTA VALLE HIGH SCHOOL   
4  01M539  NEW EXPLORATIONS INTO SCIENCE, TECHNOLOGY AND MATH HIGH SCHOOL   

  Num of SAT Test Takers  SAT Critical Reading Avg. Score  \
0                     29                            355.0   
1                     91                            383.0   
2                     70                            377.0   
3                     44                            390.0   
4                    159                            522.0   

   SAT Math Avg. Score  SAT Writing Avg. Score  sat_score  \
0                404.0                   363.0     1122.0   
1                423.0                   366.0     1172.0   
2                402.0                   370.0     1149.0   
3                433.0                   384.0     1207.0   
4                574.0                   525.0     1621.0   

                       SchoolName  AP Test Takers   Total Exams Taken  \
0                             NaN              NaN                NaN   
1    UNIVERSITY NEIGHBORHOOD H.S.             39.0               49.0   
2          EAST SIDE COMMUNITY HS             19.0               21.0   
3                             NaN              NaN                NaN   
4  NEW EXPLORATIONS SCI,TECH,MATH            255.0              377.0   

   Number of Exams with scores 3 4 or 5   Demographic  \
0                                   NaN  Total Cohort   
1                                  10.0  Total Cohort   
2                                   NaN  Total Cohort   
3                                   NaN  Total Cohort   
4                                 191.0  Total Cohort   

                             School Name Cohort  Total Cohort Total Grads - n  \
0  HENRY STREET SCHOOL FOR INTERNATIONAL   2006          78.0              43   
1    UNIVERSITY NEIGHBORHOOD HIGH SCHOOL   2006         124.0              53   
2             EAST SIDE COMMUNITY SCHOOL   2006          90.0              70   
3                MARTA VALLE HIGH SCHOOL   2006          84.0              47   
4  NEW EXPLORATIONS INTO SCIENCE  TECHNO   2006          46.0              46   

  Total Grads - % of cohort Total Regents - n Total Regents - % of cohort  \
0                     55.1%                36                       46.2%   
1                     42.7%                42                       33.9%   
2                     77.8%                67         74.400000000000006%   
3                       56%                40                       47.6%   
4                      100%                46                        100%   

  Total Regents - % of grads Advanced Regents - n  \
0                      83.7%                    0   
1                      79.2%                    8   
2                      95.7%                    0   
3                      85.1%                   17   
4                       100%                   31   

  Advanced Regents - % of cohort Advanced Regents - % of grads  \
0                             0%                            0%   
1                           6.5%                         15.1%   
2                             0%                            0%   
3                          20.2%           36.200000000000003%   
4            67.400000000000006%           67.400000000000006%   

  Regents w/o Advanced - n Regents w/o Advanced - % of cohort  \
0                       36                              46.2%   
1                       34                              27.4%   
2                       67                74.400000000000006%   
3                       23                              27.4%   
4                       15          

In [38]:
combined.shape

(363, 159)

## 10. Filling in missing values

the inner joins resulted in  `116`  fewer rows in  `sat_results` . This is because pandas couldn't find the  `DBN`  values that existed in  `sat_results`  in the other data sets. While this is worth investigating, we're currently looking for high-level correlations, so we don't need to dive into which  `DBNs`  are missing.

we now have many columns with null ( `NaN` ) values. This is because we chose to do  `left`  joins, where some columns may not have had data. The data set also had some missing values to begin with. If we hadn't performed a  `left`  join, all of the rows with missing data would have been lost in the merge process, which wouldn't have left us with many high schools in our data set.

There are several ways to handle missing data, and we'll cover them in more detail later on. For now, we'll just fill in the missing values with the overall mean for the column

In [39]:
combined = combined.fillna(combined.mean())
combined = combined.fillna(0)
combined.head()

DBN                                                     SCHOOL NAME  \
0  01M292                   HENRY STREET SCHOOL FOR INTERNATIONAL STUDIES   
1  01M448                             UNIVERSITY NEIGHBORHOOD HIGH SCHOOL   
2  01M450                                      EAST SIDE COMMUNITY SCHOOL   
3  01M509                                         MARTA VALLE HIGH SCHOOL   
4  01M539  NEW EXPLORATIONS INTO SCIENCE, TECHNOLOGY AND MATH HIGH SCHOOL   

  Num of SAT Test Takers  SAT Critical Reading Avg. Score  \
0                     29                            355.0   
1                     91                            383.0   
2                     70                            377.0   
3                     44                            390.0   
4                    159                            522.0   

   SAT Math Avg. Score  SAT Writing Avg. Score  sat_score  \
0                404.0                   363.0     1122.0   
1                423.0                   366.0     1172.0   
2                402.0                   370.0     1149.0   
3                433.0                   384.0     1207.0   
4                574.0                   525.0     1621.0   

                       SchoolName  AP Test Takers   Total Exams Taken  \
0                               0       129.028846         197.038462   
1    UNIVERSITY NEIGHBORHOOD H.S.        39.000000          49.000000   
2          EAST SIDE COMMUNITY HS        19.000000          21.000000   
3                               0       129.028846         197.038462   
4  NEW EXPLORATIONS SCI,TECH,MATH       255.000000         377.000000   

   Number of Exams with scores 3 4 or 5   Demographic  \
0                                153.45  Total Cohort   
1                                 10.00  Total Cohort   
2                                153.45  Total Cohort   
3                                153.45  Total Cohort   
4                                191.00  Total Cohort   

                             School Name Cohort  Total Cohort Total Grads - n  \
0  HENRY STREET SCHOOL FOR INTERNATIONAL   2006          78.0              43   
1    UNIVERSITY NEIGHBORHOOD HIGH SCHOOL   2006         124.0              53   
2             EAST SIDE COMMUNITY SCHOOL   2006          90.0              70   
3                MARTA VALLE HIGH SCHOOL   2006          84.0              47   
4  NEW EXPLORATIONS INTO SCIENCE  TECHNO   2006          46.0              46   

  Total Grads - % of cohort Total Regents - n Total Regents - % of cohort  \
0                     55.1%                36                       46.2%   
1                     42.7%                42                       33.9%   
2                     77.8%                67         74.400000000000006%   
3                       56%                40                       47.6%   
4                      100%                46                        100%   

  Total Regents - % of grads Advanced Regents - n  \
0                      83.7%                    0   
1                      79.2%                    8   
2                      95.7%                    0   
3                      85.1%                   17   
4                       100%                   31   

  Advanced Regents - % of cohort Advanced Regents - % of grads  \
0                             0%                            0%   
1                           6.5%                         15.1%   
2                             0%                            0%   
3                          20.2%           36.200000000000003%   
4            67.400000000000006%           67.400000000000006%   

  Regents w/o Advanced - n Regents w/o Advanced - % of cohort  \
0                       36                              46.2%   
1                       34                              27.4%   
2                       67                74.400000000000006%   
3                       23                              27.4%   
4                       15          

## 11. Adding a school district column for mapping

We've finished cleaning and combining our data! We now have a clean data set on which we can base our analysis. Mapping the statistics out on a school district level might be an interesting way to analyze them. Adding a column to the data set that specifies the school district will help us accomplish this.

The school district is just the first two characters of the  `DBN` . We can apply a function over the  `DBN`  column of  `combined`  that pulls out the first two letters.

In [40]:
combined['school_dist'] = combined.DBN.apply(lambda x: x[:2])
combined.school_dist.head()

0    01
1    01
2    01
3    01
4    01
Name: school_dist, dtype: object

We now have a clean data set we can analyze! We've done a lot. We've gone from having several messy sources to one clean, combined, data set that's ready for analysis.

Along the way, we've learned about:

* How to handle missing values
* Different types of merges
* How to condense data sets
* How to compute averages across dataframes

Data scientists rarely start out with tidy data sets, which makes cleaning and combining them one of the most critical skills any data professional can learn.

we'll analyze our clean data to find correlations and create maps.